rfd/0141-tsh-bench-db.md

```diff
@@ -0,0 +1,279 @@
+
+The "root" command for each protocol will execute the following:
+* Starts a connection with the database (using local proxy);
+* Issue a ping-like query to ensure the connection is working. This also covers
+  drivers that connect only when queries are executed.
+* Closes the connection.
+
+User certificate generation and local proxy start will not affect the final
+results. The connections will be issued to a single local proxy, reducing the
+resources used locally.
+
+The metrics generated by those subcommands will be restricted to raw end-to-end
+timings (without any breakdown). It must be used alongside backend metrics to
+have full detailed information on the service behavior.
+
+```mermaid
+sequenceDiagram
+  participant tsh
+  participant Teleport
+  participant DB as Database
+
+  tsh->>Teleport: Generate Certificates (tsh db login)
+
+  loop every execution
+    tsh->>Teleport: Establish connection
+    Teleport->>DB: Connect
+    DB->>tsh: Connection
+    tsh->>DB: Ping
+  end
+```
+
+### Commands
+Both PostgreSQL and MySQL will have the same command arguments and flags.
+
+```code
+tsh bench postgres [--db-user=] [--db-name=] [--uri] [database]
+```
+
+```code
+tsh bench mysql [--db-user=] [--db-name=] [--uri] [database]
+```
+
+| Flag        | Description |
+| ----------- | ----------- |
+| `--db-user` | Database user used to connect to the target database. The user must have enough permissions on the database to execute all the benchmark queries. |
+| `--db-name` | Database name where benchmark queries will be executed. Depending on the database protocol, this isn't required. |
+| `--uri`     | Direct database access URI. When provided, the benchmark will issue connections directly to this database, and no Teleport is involved in the testing. It must contain all the connection information, including authentication credentials. The contents from --db-user and --db-name will be ignored when provided. |
+| `database`  | Teleport target database name. Available databases can be retrieved by running `tsh db ls`. If not provided, the `--uri` flag (for direct database connection) must be present. | 
+
+Note: The `--rate` common flag will indicate how many connections per second will be
+established.
+
+**Examples:**
+```bash
+# Executes 25 connections per second during 10 seconds without errors but also
+# without meeting the target requests.
+$ tsh bench postgres --rate=25 --duration=10s --db-user=postgres --db-name=postgres postgres-dev
+
+* Requests originated: 88
+* Requests failed: 0
+
+Histogram
+
+Percentile Response Duration
+---------- -----------------
+25         1346 ms
+50         3535 ms
+75         5259 ms
+90         5847 ms
+95         6663 ms
+99         7271 ms
+100        7311 ms
+
+# Executes 100 connections per second during 30 seconds with errors (only the
+# last error is shown).
+$ tsh bench postgres --rate 100 --duration 30s --db-user=postgres --db-name=postgres postgres-dev
```

Comment:
The commands will support exporting the latency profile like "regular" `tsh bench` does, correct?

Reply:
Yes, it does. (I will add an example with this option).